From 6bc86730c441dcc7d9fbd01bf773b454e3505e72 Mon Sep 17 00:00:00 2001
From: Pekka Vuorela <pekka.vuorela@jollamobile.com>
Date: Fri, 13 Oct 2017 15:25:51 +0300
Subject: [PATCH 14/38] zypp Backend: Download progress reporting

Include downloading of packages in total progress calculation.

[backends] zypp: Overall download progress reporting
---
 backends/zypp/pk-backend-zypp.cpp | 28 ++++++++++++++++++++++++----
 1 file changed, 24 insertions(+), 4 deletions(-)

diff --git a/backends/zypp/pk-backend-zypp.cpp b/backends/zypp/pk-backend-zypp.cpp
index 5d08d4423..44dd11868 100644
--- a/backends/zypp/pk-backend-zypp.cpp
+++ b/backends/zypp/pk-backend-zypp.cpp
@@ -211,6 +211,7 @@ static PkBackendZYppPrivate *priv = 0;
 
 // FIXME: should merge with _dl_progress / _dl_count
 /* Overall progress update helpers */
+void zypp_backend_download_finished(PkBackendJob *job);
 void zypp_backend_installation_finished(PkBackendJob *job);
 void zypp_backend_removal_finished(PkBackendJob *job);
 
@@ -433,6 +434,7 @@ struct DownloadProgressReportReceiver : public zypp::callback::ReceiveReport<zyp
 	{
 		MIL << resolvable << " " << error << " " << _package_id << std::endl;
 		update_sub_percentage (100, PK_STATUS_ENUM_DOWNLOAD);
+		zypp_backend_download_finished(_job);
 		// FIXME: uncomment when ExecCounter and _dl_progress are unified
 		//pk_backend_job_set_percentage(_job, (double)++_dl_progress / _dl_count * 100);
 		clear_package_id ();
@@ -581,13 +583,15 @@ struct ExecCounters {
 		, current_installs(0)
 		, total_removals(0)
 		, current_removals(0)
+		, total_downloads(0)
+		, current_downloads(0)
 	{
 	}
 
 	void update(PkBackendJob *job)
 	{
-		int total = (total_installs + total_removals);
-		int current = (current_installs + current_removals);
+		int total = (total_downloads + total_installs + total_removals);
+		int current = (current_downloads + current_installs + current_removals);
 
 		if (current > total) {
 			MIL << "current > total!" << std::endl;
@@ -604,12 +608,16 @@ struct ExecCounters {
 		current_installs = 0;
 		total_removals = 0;
 		current_removals = 0;
+		total_downloads = 0;
+		current_downloads = 0;
 	}
 
 	int total_installs;
 	int current_installs;
 	int total_removals;
 	int current_removals;
+	int total_downloads;
+	int current_downloads;
 };
 
 class PkBackendZYppPrivate {
@@ -622,6 +630,12 @@ class PkBackendZYppPrivate {
 	ExecCounters exec;
 };
 
+void zypp_backend_download_finished(PkBackendJob *job)
+{
+	priv->exec.current_downloads += 1;
+	priv->exec.update(job);
+}
+
 void zypp_backend_installation_finished(PkBackendJob *job)
 {
 	priv->exec.current_installs += 1;
@@ -1632,13 +1646,19 @@ zypp_perform_execution (PkBackendJob *job, ZYpp::Ptr zypp, PerformType type, gbo
 			}
 
 			if (it->status().isToBeInstalled()) {
-				priv->exec.total_installs += 1;
+				if (!only_download) {
+					priv->exec.total_installs += 1;
+				}
+				priv->exec.total_downloads += 1;
 			} else if (it->status().isToBeUninstalled() &&
 					!it->status().isToBeUninstalledDueToUpgrade()) {
-				priv->exec.total_removals += 1;
+				if (!only_download) {
+					priv->exec.total_removals += 1;
+				}
 			}
 		}
 		MIL << "Summary before commit: " << std::endl;
+		MIL << " total downloads = " << priv->exec.total_downloads << std::endl;
 		MIL << " total installs = " << priv->exec.total_installs << std::endl;
 		MIL << " total removals = " << priv->exec.total_removals << std::endl;
 
-- 
2.13.6

